datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -668,16 +670,40 @@ impl OptimizerRule for PushDownFilter {
 
                         (field.qualified_name(), expr)
                     })
-                    .collect::<HashMap<_, _>>();
+                    .partition(|(_, value)| matches!(value, Expr::ScalarFunction(f) if f.fun.volatility() == Volatility::Volatile));
 
-                // re-write all filters based on this projection
-                // E.g. in `Filter: b\n  Projection: a > 1 as b`, we can swap them, but the filter must be "a > 1"
-                let new_filter = LogicalPlan::Filter(Filter::try_new(
-                    replace_cols_by_name(filter.predicate.clone(), &replace_map)?,
-                    projection.input.clone(),
-                )?);
+                let mut push_predicates = vec![];
+                let mut keep_predicates = vec![];
+                for expr in split_conjunction_owned(filter.predicate.clone()).into_iter()
+                {
+                    if contain(expr.clone(), &volatile_map.clone())? {
+                        keep_predicates.push(expr);
+                    } else {
+                        push_predicates.push(expr);
+                    }
+                }
 
-                child_plan.with_new_inputs(&[new_filter])?
+                match conjunction(push_predicates) {
+                    Some(expr) => {
+                        // re-write all filters based on this projection
+                        // E.g. in `Filter: b\n  Projection: a > 1 as b`, we can swap them, but the filter must be "a > 1"
+                        let new_filter = LogicalPlan::Filter(Filter::try_new(
+                            replace_cols_by_name(expr, &non_volatile_map)?,
+                            projection.input.clone(),
+                        )?);
+
+                        if keep_predicates.is_empty() {
+                            child_plan.with_new_inputs(&[new_filter])?
+                        } else {
+                            let child_plan = child_plan.with_new_inputs(&[new_filter])?;
+                            LogicalPlan::Filter(Filter::try_new(
+                                conjunction(keep_predicates).unwrap(),
+                                Arc::new(child_plan),
+                            )?)
+                        }
```

Comment:
You maybe could avoid the unwrap with something like this instead:

```suggestion
                        match conjunction(keep_predicates) {
                          None =>  child_plan.with_new_inputs(&[new_filter])?
                          Some(keep_predicate) => {
                            let child_plan = child_plan.with_new_inputs(&[new_filter])?;
                            LogicalPlan::Filter(Filter::try_new(
                                keep_predicate
                                Arc::new(child_plan),
                            )?)
                        }
```